### PR TITLE
Adding a spider for RN Macaiba

### PR DESCRIPTION
**Pull request by phrfpeixoto**
Along with the spider, I'm also adding a new `ItemPipeline` that checks the item's `source_text` property against a spider-defined pattern, to determine whether or not that Gazette may be an Extra Issue.

### CODE REVIEW

File: processing/data_collection/gazette/pipelines.py [hunk 1]
```diff
@@ -22,6 +20,30 @@ def process_item(self, item, spider):
...
+        # Avoid touching an item that was manually defined.
+        if "is_extra_edition" in item:
+            return item
+
+        # No source_text, nothing to do
+        if item["source_text"] is None:
+            return item
+
+        # If the spider has not defined a pattern, go away
+        if not hasattr(spider, "EXTRA_ISSUE_PATTERN"):
```
Comment (ejulio) [inline processing/data_collection/gazette/pipelines.py:39]: I'd put this check before the item content

File: processing/data_collection/gazette/pipelines.py [hunk 2]
```diff
@@ -22,6 +20,30 @@ def process_item(self, item, spider):
...
+            return item
+
+        if re.findall(spider.EXTRA_ISSUE_PATTERN, item["source_text"], re.IGNORECASE):
```
Comment (ejulio) [inline processing/data_collection/gazette/pipelines.py:42]: I guess `re.search` would be better, as it stops in the first match and doesn't need to find all matches

File: processing/data_collection/gazette/spiders/rn_macaiba.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+import datetime as dt
+
+import scrapy
+
+from gazette.items import Gazette
+from gazette.spiders.base import BaseGazetteSpider
+
+
+class RnMacaiba(BaseGazetteSpider):
+    start_date: dt.date
```
Comment (ejulio) [inline processing/data_collection/gazette/spiders/rn_macaiba.py:13]: I think this one should be set by the base class, as it can be a command line argument

File: processing/data_collection/gazette/spiders/rn_macaiba.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    TERRITORY_ID = "2407104"
+    EXTRA_ISSUE_PATTERN = r"Edição Extraordinária"
+    name = "rn_macaiba"
+    allowed_domains = ["macaiba.rn.gov.br"]
+
+    def __init__(self, start_date=None, end_date=None):
+        self.start_date = dt.date.today().replace(day=1)
+        self.end_date = dt.date.today()
+
+        if start_date is not None:
```
Comment (ejulio) [inline processing/data_collection/gazette/spiders/rn_macaiba.py:25]: Regarding my comment above, I think this is done in base `__init__`

File: processing/data_collection/gazette/spiders/rn_macaiba.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+            parsed_data = dateparser.parse(start_date)
+            if parsed_data is not None:
+                self.start_date = parsed_data.date()
+
+        if end_date is not None:
```
Comment (ejulio) [inline processing/data_collection/gazette/spiders/rn_macaiba.py:30]: If not done yet, maybe it should be done in the base class as well

File: processing/data_collection/gazette/spiders/rn_macaiba.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+            gazette = item.css("::text").get().strip()
+            date_match = re.findall(
+                r"(\d{2}) ([a-zç]+), (\d{4})", gazette, re.IGNORECASE
+            )
+            if len(date_match) != 1:
+                self.logger.error(f"Could not parse the date for Gazette {gazette}")
+                continue
+
+            item_date = " de ".join(date_match.pop())
+            item_date = dateparser.parse(item_date).date()
```
Comment (ejulio) [inline processing/data_collection/gazette/spiders/rn_macaiba.py:61]: I think a specific patter is better, relying on date parser can be a problem as it needs to identify the pattern and it can be a huge search through many defined patterns

File: processing/data_collection/gazette/spiders/rn_macaiba.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+            file_url = item.css("a::attr(href)").get().lower()
+
+            if not file_url.endswith("pdf"):
+                self.logger.error(
```
Comment (ejulio) [inline processing/data_collection/gazette/spiders/rn_macaiba.py:65]: Is it really an error or just `info`/`warning`?
My opinion is that errors are things we'll need to fix, warnings are things we should be careful and info is just info :stuck_out_tongue:

File: processing/data_collection/gazette/spiders/rn_macaiba.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+            method="POST",
+            formdata={
+                "inicio": self.start_date.isoformat(),
+                "fim": self.end_date.isoformat(),
+            },
+        )
+
+    def parse(self, response):
+        for item in response.css("div.conteudo li.collection-item div"):
+            gazette = item.css("::text").get().strip()
```
Comment (ejulio) [inline processing/data_collection/gazette/spiders/rn_macaiba.py:52]: Maybe you can `item.css("::text").re(regex here)`, not sure in this case

**Review by ejulio — CHANGES_REQUESTED** [review]

File: processing/data_collection/gazette/pipelines.py [hunk 9]
```diff
@@ -22,6 +20,30 @@ def process_item(self, item, spider):
         return item
 
 
+class ExtraIssuePipeline:
```
Comment (jvanz) [inline processing/data_collection/gazette/pipelines.py:23]: Please, remove this pipeline item. The file data processing is under a change. We will move the processing pipeline for another repository. It's the same that I sent you in another PR.

Comment (jvanz) [reply processing/data_collection/gazette/pipelines.py:23]: For now, if you can identify the type from the web page, url or file name would be nice.

File: processing/data_collection/gazette/spiders/rn_macaiba.py (hunk 3 above)
Comment (jvanz) [reply processing/data_collection/gazette/spiders/rn_macaiba.py:13]: Yes, that's right.

**Review by jvanz — CHANGES_REQUESTED** [review]

File: processing/data_collection/gazette/pipelines.py (hunk 9 above)
Comment (phrfpeixoto) [reply processing/data_collection/gazette/pipelines.py:23]: @jvanz That's the only way I found to identify it. There's no information on the page, URL or filename. Do you still want me to remove it?

File: processing/data_collection/gazette/spiders/rn_macaiba.py (hunk 6 above)
Comment (phrfpeixoto) [reply processing/data_collection/gazette/spiders/rn_macaiba.py:61]: I've chosen to use this simply because it's already in use within the project.
Also, the month names within the page are in Portuguese, so I didn't want to mess with locales for date parsing.

Let me know your thoughts on this.

File: processing/data_collection/gazette/pipelines.py (hunk 1 above)
Comment (phrfpeixoto) [reply processing/data_collection/gazette/pipelines.py]: Done.

File: processing/data_collection/gazette/spiders/rn_macaiba.py (hunk 3 above)
Comment (phrfpeixoto) [reply processing/data_collection/gazette/spiders/rn_macaiba.py]: Done.

File: processing/data_collection/gazette/spiders/rn_macaiba.py (hunk 5 above)
Comment (phrfpeixoto) [reply processing/data_collection/gazette/spiders/rn_macaiba.py]: Done.

File: processing/data_collection/gazette/spiders/rn_macaiba.py (hunk 4 above)
Comment (phrfpeixoto) [reply processing/data_collection/gazette/spiders/rn_macaiba.py]: Done

File: processing/data_collection/gazette/pipelines.py (hunk 2 above)
Comment (phrfpeixoto) [reply processing/data_collection/gazette/pipelines.py]: Done.

File: processing/data_collection/gazette/spiders/rn_macaiba.py (hunk 7 above)
Comment (phrfpeixoto) [reply processing/data_collection/gazette/spiders/rn_macaiba.py]: Done.

File: processing/data_collection/gazette/spiders/rn_macaiba.py (hunk 8 above)
Comment (phrfpeixoto) [reply processing/data_collection/gazette/spiders/rn_macaiba.py:42]: I can't use `re.IGNORECASE` on `SelectorList.re`

File: processing/data_collection/gazette/pipelines.py (hunk 9 above)
Comment (phrfpeixoto) [reply processing/data_collection/gazette/pipelines.py:23]: @rennerocha @ejulio What do you think?

File: processing/data_collection/gazette/spiders/rn_macaiba.py (hunk 6 above)
Comment (ejulio) [reply processing/data_collection/gazette/spiders/rn_macaiba.py:51]: That's fine, just constrain the language/locale `pt-br` then.
https://dateparser.readthedocs.io/en/latest/#dateparser.parse

File: processing/data_collection/gazette/pipelines.py (hunk 9 above)
Comment (ejulio) [reply processing/data_collection/gazette/pipelines.py:23]: If we're moving the processing out of this repository, I guess there isn't much to be done.
For sure we can keep the flag in the item, to avoid running the regex later.

File: processing/data_collection/gazette/spiders/rn_macaiba.py (hunk 6 above)
Comment (phrfpeixoto) [reply processing/data_collection/gazette/spiders/rn_macaiba.py]: Done. I've restricted it to `pt` language, since the dateparser does not support `pt-BR`: https://github.com/scrapinghub/dateparser/issues/377

File: processing/data_collection/gazette/pipelines.py (hunk 9 above)
Comment (phrfpeixoto) [reply processing/data_collection/gazette/pipelines.py:23]: Do you mean keep the flag as `False`?

Comment (jvanz) [reply processing/data_collection/gazette/pipelines.py:23]: Yes, if you do not have other way to identify it. Keep the value false.

**Review by jvanz — CHANGES_REQUESTED** [review]

File: processing/data_collection/gazette/pipelines.py (hunk 9 above)
Comment (phrfpeixoto) [reply processing/data_collection/gazette/pipelines.py]: Done.

File: processing/data_collection/gazette/spiders/rn_macaiba.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+
+class RnMacaiba(BaseGazetteSpider):
+
+    TERRITORY_ID = "2407104"
+    name = "rn_macaiba"
+    allowed_domains = ["macaiba.rn.gov.br"]
+
+    def __init__(self, start_date=None, end_date=None):
+        self.start_date = dt.date.today().replace(day=1)
```
Comment (jvanz) [inline processing/data_collection/gazette/spiders/rn_macaiba.py:19]: Please, consider the `start_date` the first date with gazettes available. This is the default in other spiders.

**Review by jvanz — CHANGES_REQUESTED** [review]

File: processing/data_collection/gazette/spiders/rn_macaiba.py (hunk 10 above)
Comment (phrfpeixoto) [reply processing/data_collection/gazette/spiders/rn_macaiba.py]: Done.

**Review by ejulio — APPROVED** [review]

File: processing/data_collection/gazette/settings.py [hunk 11]
```diff
@@ -8,4 +8,4 @@
     "gazette.pipelines.ExtractTextPipeline": 200,
 }
 FILES_STORE = "/mnt/data/"
-QUERIDODIARIO_EXTRACT_TEXT_FROM_FILE = True
\ No newline at end of file
+QUERIDODIARIO_EXTRACT_TEXT_FROM_FILE = True
```
Comment (alexandrevicenzi) [inline processing/data_collection/gazette/settings.py:11]: this same change is in another of your PRs, it will cause conflicts once that one is merged

File: processing/data_collection/gazette/pipelines.py [hunk 12]
```diff
@@ -1,5 +1,5 @@
+import re
```
Comment (alexandrevicenzi) [inline processing/data_collection/gazette/pipelines.py:1]: these changes look like unrelated, or am I missing something here?

**Review by alexandrevicenzi — CHANGES_REQUESTED** [review]

File: processing/data_collection/gazette/settings.py (hunk 11 above)
Comment (phrfpeixoto) [reply processing/data_collection/gazette/settings.py:11]: Yes, but If I remove it, the automated build will fail.

File: processing/data_collection/gazette/pipelines.py (hunk 12 above)
Comment (phrfpeixoto) [reply processing/data_collection/gazette/pipelines.py]: Yes, that was left over from the middleware that was removed.
I reverted the file.

File: processing/data_collection/gazette/settings.py (hunk 11 above)
Comment (alexandrevicenzi) [reply processing/data_collection/gazette/settings.py:11]: It will not fail once one is merged, so after you can rebase it, maybe just wait a bit to get one merged and then fix the others.

Comment (phrfpeixoto) [reply processing/data_collection/gazette/settings.py:11]: That was the plan. Have the first one merged, then rebase the others. Can we start with this one?

File: processing/data_collection/gazette/spiders/rn_macaiba.py [hunk 13]
```diff
@@ -0,0 +1,66 @@
+import re
+
+import dateparser
+import datetime as dt
+
+import scrapy
+
+from gazette.items import Gazette
+from gazette.spiders.base import BaseGazetteSpider
```
Comment (jvanz) [inline processing/data_collection/gazette/spiders/rn_macaiba.py:9]:
```suggestion
import datetime as dt
import re

import dateparser
import scrapy

from gazette.items import Gazette
from gazette.spiders.base import BaseGazetteSpider
```

**Review by jvanz — CHANGES_REQUESTED** [review]

**Review by jvanz — APPROVED** [review]

**Comment by jvanz** [comment]
@phrfpeixoto, can you rebase the branch and fix the conflicts?

**Comment by phrfpeixoto** [comment]
@jvanz Done.

File: processing/data_collection/gazette/spiders/rn_macaiba.py (hunk 13 above)
Comment (phrfpeixoto) [reply processing/data_collection/gazette/spiders/rn_macaiba.py]: Done.

**Comment by phrfpeixoto** [comment]
@jvanz Can we merge this?

File: processing/data_collection/gazette/spiders/base.py [hunk 14]
```diff
@@ -26,8 +29,16 @@ def __init__(self, start_date=None, *args, **kwargs):
                     f"Unable to parse {start_date}. Use %Y-%m-d date format."
                 )
                 raise
-        else:
-            self.logger.info("Collecting all gazettes available")
+
+        if end_date is not None:
+            try:
+                self.end_date = datetime.strptime(end_date, "%Y-%m-%d").date()
+                self.logger.info(f"Collecting gazettes until {self.end_date}")
```
Comment (jvanz) [inline processing/data_collection/gazette/spiders/base.py:36]: Move this log to the end of the `__init__` method. Thus, it will always be printed. ;)

Comment (jvanz) [reply processing/data_collection/gazette/spiders/base.py:36]: Please, do the same if the log statement for the start_date

**Review by jvanz — CHANGES_REQUESTED** [review]

**Comment by phrfpeixoto** [comment]
Is this still relevant?

**Comment by rennerocha** [comment]
Website changed, so a new spider need to be developed and unfortunately nothing can be reused from here.
This city will be developed in https://github.com/okfn-brasil/querido-diario/pull/601
